xarray/core/dataset.py

```diff
@@ -1077,12 +1077,41 @@ def relevant_keys(mapping):
         if not utils.is_scalar(dim) and not isinstance(dim, DataArray):
             dim = as_variable(dim, name='points')
 
+        indexers_dict = dict(indexers)
+        non_indexed = list(set(self.dims) - indexer_dims)
+
+        # need to generate a new dimension
+        points_len = lengths.pop()
+
+        variables = OrderedDict()
+
+        # add : slices for the non-indexed dimensions
+        slc = [indexers_dict[k] if k in indexers_dict else slice(None, None) for k in self.dims]
+        coords = [self.variables[k] for k in non_indexed]
+
+        # TODO need to figure out how to make sure we get the indexed vs non indexed dimensions in the right order
```

Comment:
If you transpose the arrays so that the dimensions being indexed are always leading dimensions (something like `reordered = self.transpose(*(indexed_dims + non_indexed_dims))`), indexing `.vindex` on a dask array and directly on a numpy array will give the same result.

You really don't want to allow indexed dimensions in the middle somewhere because NumPy has unintuitive results for reordering axes in some edge cases.